proposals/0000-existentials.rst

```diff
@@ -0,0 +1,587 @@
+   is lazy.
+
+   Accordingly, if there are multiple givens arising from ``/\`` that could solve a wanted,
+   we choose none of them. If there are no other ways of solving the wanted, then the
+   user's program is rejected. Alternatively, we could work this new flavor of given
+   into the scheme outlined in `#17934`_.
+
+#. **Compilation.** The new ``Pack`` and ``Open`` constructs can be erased entirely
+   during code generation. This is why the representation of ``exists (a :: k). ty``
+   is the same as the representation of ``ty``. This is exactly the same as the treatment
+   of ``forall`` and type application.
+
+Optional Extension
+------------------
+
+Extend the treatment of existentials to work in types. (Do *not* extend the treatment
+of constraints with ``/\``.) This would affect the kind
+inference algorithm analogously to the effects on the type inference algorithm above.
+The only complication is that we would need ``TWitness`` instead of ``Witness``.
+A ``TWitness`` accepts a type, not an expression. (In the glorious future of Dependent
+Haskell, ``TWitness`` becomes a synonym for ``Witness``.)
+
+Supporting this feature would require yet more extensions to Core:
+* ``PackTy`` and ``OpenTy`` in ``Type``, analogous to ``Pack`` and ``Open`` in ``Expr``.
+* ``TWitness`` in ``Type``, analogous to ``Witness``.
+* ``TWitnessPackCo`` and ``TWitnessCo`` in ``Coercion``, analogous to ``WitnessPackCo`` and ``WitnessCo``.
+
+I would prefer not to have this extension, but save type-level support for packing
+and opening existentials until we have proper dependent types.
+
+Examples
+--------
+
+1. See the filter_ example, above.
+
+#. ::
+
+     toVec :: [a] -> exists n. KnownNat n /\ Vec n a
+     toVec []     = VNil
+     toVec (x:xs) = x :> toVec xs
+
+     -- NB: requires inputs to have the same length
+     zipWithV :: (a -> b -> c) -> Vec n a -> Vec n b -> Vec n c
+     zipWithV = ...
+
+     -- from GHC.TypeLits:
+     sameNat :: forall a b proxy1 proxy2. (KnownNat a, KnownNat b) => proxy1 a -> proxy2 b -> Maybe (a :~: b)
+
+     -- returns first argument iff its length matches that of the second, in O(1) time
+     sameLengthAs :: forall m n a b. (KnownNat m, KnownNat n) => Vec m a -> Vec n b -> Maybe (Vec n a)
+     sameLengthAs v1 _ = do Refl <- sameNat @m @n Proxy Proxy
+                            return v1
+
+     frob :: (a -> b -> c) -> [a] -> [b] -> exists n. Maybe (Vec n c)
+     frob f as bs = do let as' = toVec as
+                           bs' = toVec bs
+                       as'' <- sameLengthAs as' bs'
+                       return (zipWithV f as'' bs')
+
+#. Imagine this within GHC::
+
+     -- in GHC.Utils.Outputable
+     sep :: [SDoc] -> SDoc
+     ppr :: Outputable a => a -> SDoc
+
+     pprs :: [exists a. Outputable a /\ a] -> SDoc
+     pprs = sep . map ppr    -- this type-checks because of the rules for class constraints on existentials
```

Comment:
I'm a bit surprised this would work, and probably I'm reading something wrong. The parameter seems to be a list of possibly different outputable types (else the exists would be outside the `[]`), together with their witnesses. So operationally some sort of pair? But `ppr`, in core, takes two arguments. And `map` doesn't uncurry `ppr` itself. Will the desugarer wrap the call to `ppr` in some form of `uncurryC` here?

Reply:
Blargh. Answering your question killed my special rules around inferring e.g. `[W] C (exists a. ty)`, because those rules were bogus. I've left them as future work.